velox/functions/prestosql/JsonFunctions.cpp

```diff
@@ -130,45 +224,80 @@ class JsonParseFunction : public exec::VectorFunction {
   }
 
  private:
-  simdjson::error_code parse(size_t size) const {
+  simdjson::error_code parse(size_t size, JsonViewPtr& jsonView) const {
     simdjson::padded_string_view paddedInput(
         paddedInput_.data(), size, paddedInput_.size());
     SIMDJSON_ASSIGN_OR_RAISE(auto doc, simdjsonParse(paddedInput));
-    SIMDJSON_TRY(validate<simdjson::ondemand::document&>(doc));
+    SIMDJSON_TRY(validate<simdjson::ondemand::document&>(doc, jsonView));
     if (!doc.at_end()) {
       return simdjson::TRAILING_CONTENT;
     }
     return simdjson::SUCCESS;
   }
 
   template <typename T>
-  static simdjson::error_code validate(T value) {
+  static simdjson::error_code validate(T value, JsonViewPtr& jsonView) {
     SIMDJSON_ASSIGN_OR_RAISE(auto type, value.type());
     switch (type) {
       case simdjson::ondemand::json_type::array: {
         SIMDJSON_ASSIGN_OR_RAISE(auto array, value.get_array());
+
+        std::vector<JsonViewPtr> arrayPtr;
         for (auto elementOrError : array) {
           SIMDJSON_ASSIGN_OR_RAISE(auto element, elementOrError);
-          SIMDJSON_TRY(validate(element));
+          JsonViewPtr elementPtr;
+          SIMDJSON_TRY(validate(element, elementPtr));
+          arrayPtr.push_back(elementPtr);
         }
+
+        jsonView = std::make_shared<JsonArrayView>(arrayPtr);
```

Comment:
When looking at the constructor of JsonArrayView, we are copying the std::vector<JsonViewPtr> twice. What is the conventional way to do in this case?

Reply:
In almost all cases the compiler, I think the compiler will end up doing copy elision, so this should not create a copy as these types are not marked volatile . However to further satisfy you , i can do an explicit std::move.